src/software/ai/passing/receiver_position_generator.hpp

```diff
@@ -0,0 +1,285 @@
+    const auto &receiver_config = passing_config_.receiver_position_generator_config();
+
+    // Add the previous best sampled receiving positions
+    for (const auto &[zone, prev_best_receiving_position] : prev_best_receiving_positions)
+    {
+        Pass pass = Pass::fromDestReceiveSpeed(pass_origin, prev_best_receiving_position,
+                                               passing_config_);
+        // Increase the rating of the previous best receiving positions to
+        // discourage changing the receiver positions too much.
+        double receiver_position_rating =
+            rateReceivingPosition(world, pass, passing_config_) *
+            receiver_config.previous_best_receiver_position_score_multiplier();
+        best_receiving_positions.insert_or_assign(
+            zone, PassWithRating{pass, receiver_position_rating});
+    }
+
+    auto all_zones = pitch_division_->getAllZoneIds();
+
+    // Begin by sampling a few passes per zone to get an initial estimate of the best
+    // receiving zones
+    updateBestReceiverPositions(world, pass_origin, all_zones,
+                                receiver_config.num_initial_samples_per_zone());
+
+    // Sort the zones based on initial ratings
+    auto zone_comparator = [&](const ZoneEnum &z1, const ZoneEnum &z2) {
+        return best_receiving_positions.find(z1)->second.rating >
+               best_receiving_positions.find(z2)->second.rating;
+    };
+    std::sort(all_zones.begin(), all_zones.end(), zone_comparator);
+
+    const Angle min_angle_diff_between_receivers =
+        Angle::fromDegrees(receiver_config.min_angle_between_receivers_deg());
+    std::vector<ZoneEnum> top_zones;
+    for (unsigned int i = 0; i < all_zones.size() && top_zones.size() < num_positions;
+         i++)
+    {
+        // Only add zones that are not too close to the previous top receiver positions
+        // to encourage spreading out the receivers
+        Angle curr_pass_angle =
+            best_receiving_positions.find(all_zones[i])->second.pass.passerOrientation();
+        bool no_prev_receivers_close =
+            std::none_of(top_zones.begin(), top_zones.end(), [&](const ZoneEnum &zone) {
+                return curr_pass_angle.minDiff(best_receiving_positions.find(zone)
+                                                   ->second.pass.passerOrientation()) <
+                       min_angle_diff_between_receivers;
+            });
+        no_prev_receivers_close =
+            no_prev_receivers_close &&
+            std::none_of(
+                existing_receiver_positions.begin(), existing_receiver_positions.end(),
+                [&](const Point &existing_receiver_position) {
+                    return curr_pass_angle.minDiff(
+                               (pass_origin - existing_receiver_position).orientation()) <
+                           min_angle_diff_between_receivers;
+                });
```

Comment:
this logic is a bit hard to follow -- maybe refactoring into a private function is better